### ISSUE TITLE
`Summa` Smart Contract deals with time stamps 

### ISSUE THREAD

**Issue by enricobottazzi** [summa-dev/summa-solvency#112]
- Add `timestamp` which indicates the moment in which the merkle sum tree of liabilities was created
- Add `block_hashes` which indicates the moment in which the proof of assets is taken. It should be the next available block hash at `timestamp`

**Comment by alxkzmn** [comment]
Normally, this is not done on blockchain because this is metadata available for any blockchain indexer. I.e., when frontends/backends work with the emitted contract events, they can easily read their timestamps (or block hashes). Do we need this info in the contract for future computations or for other contracts to read?

**Comment by enricobottazzi** [comment]
Let me rephrase it since the issue was poorly explained. 

The origin of the issue is that we need to make sure that the snapshot for assets and liabilities is taken at the same time.

The way it works right now is that the liabilities are snapshotted at the moment in which we create the csv file containing the list of users and their balances. The assets instead are fetched by the contract at the current blockhash. So this creates a misalignement. An attendee of my presentation in Paris also recognized this issue.

Ideally, to solve this we can add to the π of Solvency a UNIX `liabilities_timestamp` that indicates the time in which the csv file for the liabilities was snapshotted.

When the proof gets sent for verification to the Summa Smart Contract the various balances are not fetched at the current block but at the very next block available at `liabilities_timestamp`.

https://github.com/summa-dev/summa-solvency/blob/ded7155c585eabef653bfc84ec26ca10a6acd632/contracts/src/Summa.sol#L96-L97

**Comment by alxkzmn** [comment]
Let me translate it into a new discussed architecture:
```
/**
 * @dev Interface for asset balance retriever contracts
 */
interface IBalanceRetriever {
    /**
     * @dev Get the balance of an address for a given asset type. Gets address balance for ETH, balanceOf() for ERC20, or calls an oracle for non-native assets
     * @param _address The address to check the balance of
     * @param args Additional arguments needed to get the balance (e.g., in case of ERC20, the address of the ERC20 contract)
     * @return The balance of the address for the given asset type
     */
    function getAddressBalance(
        bytes memory _address,
        bytes memory args
    ) external view returns (uint256);
}
```
We need to modify this to have the timestamp as an argument (e.g., for future implementation using Axiom):
```
interface IBalanceRetriever {
    function getAddressBalance(
        bytes memory _address,
        bytes memory args,
        uint256 timestamp
    ) external view returns (uint256);
}
```
Is that correct?

**Comment by enricobottazzi** [comment]
Yep that's correct!

**Comment by alxkzmn** [comment]
Implemented in #131.
